File: csharp/Platform.Random.Tests/RandomExtensionsTests.cs
using System.Linq;
using Platform.Ranges;
using Xunit;

namespace Platform.Random.Tests
{
    public class RandomExtensionsTests
    {
        [Fact]
        public void NextUInt64_WithoutRange_ReturnsValidValues()
        {
            var random = new System.Random(42);
            var values = new ulong[100];
            
            for (var i = 0; i < values.Length; i++)
            {
                values[i] = random.NextUInt64();
                Assert.InRange(values[i], ulong.MinValue, ulong.MaxValue);
            }
            
            var uniqueValues = values.Distinct().Count();
            Assert.True(uniqueValues > 50, "Should generate varied random values");
        }

        [Fact]
        public void NextUInt64_WithRange_ReturnsValuesInRange()
        {
            var random = new System.Random(42);
            var range = new Range<ulong>(10UL, 20UL);
            
            for (var i = 0; i < 100; i++)
            {
                var value = random.NextUInt64(range);
                Assert.InRange(value, range.Minimum, range.Maximum);
            }
        }

        [Fact]
        public void NextUInt64_WithMinimumRange_ReturnsMinimumValue()
        {
            var random = new System.Random(42);
            var range = new Range<ulong>(5UL, 5UL);
            
            var value = random.NextUInt64(range);
            Assert.Equal(5UL, value);
        }

        [Fact]
        public void NextUInt64_WithLargeRange_ReturnsValidValues()
        {
            var random = new System.Random(42);
            var range = new Range<ulong>(ulong.MaxValue - 1000UL, ulong.MaxValue);
            
            for (var i = 0; i < 10; i++)
            {
                var value = random.NextUInt64(range);
                Assert.InRange(value, range.Minimum, range.Maximum);
            }
        }

        [Fact]
        public void NextUInt64_WithZeroRange_ReturnsZero()
        {
            var random = new System.Random(42);
            var range = new Range<ulong>(0UL, 0UL);
            
            var value = random.NextUInt64(range);
            Assert.Equal(0UL, value);
        }

        [Fact]
        public void NextUInt64_WithDifferentSeeds_GeneratesDifferentSequences()
        {
            var random1 = new System.Random(1);
            var random2 = new System.Random(2);
            var range = new Range<ulong>(0UL, 1000UL);
            
            var values1 = new ulong[10];
            var values2 = new ulong[10];
            
            for (var i = 0; i < 10; i++)
            {
                values1[i] = random1.NextUInt64(range);
                values2[i] = random2.NextUInt64(range);
            }
            
            Assert.NotEqual(values1, values2);
        }

        [Fact]
        public void NextBoolean_GeneratesBothTrueAndFalse()
        {
            var random = new System.Random(42);
            var trueCount = 0;
            var falseCount = 0;
            
            for (var i = 0; i < 1000; i++)
            {
                var value = random.NextBoolean();
                if (value)
                {
                    trueCount++;
                }
                else
                {
                    falseCount++;
                }
            }
            
            Assert.True(trueCount > 0, "Should generate at least one true value");
            Assert.True(falseCount > 0, "Should generate at least one false value");
            Assert.InRange(trueCount, 300, 700);
            Assert.InRange(falseCount, 300, 700);
        }

        [Fact]
        public void NextBoolean_WithSameSeed_GeneratesSameSequence()
        {
            var random1 = new System.Random(123);
            var random2 = new System.Random(123);
            
            for (var i = 0; i < 10; i++)
            {
                var value1 = random1.NextBoolean();
                var value2 = random2.NextBoolean();
                Assert.Equal(value1, value2);
            }
        }

        [Fact]
        public void NextBoolean_WithDifferentSeeds_GeneratesDifferentSequences()
        {
            var random1 = new System.Random(1);
            var random2 = new System.Random(2);
            
            var values1 = new bool[20];
            var values2 = new bool[20];
            
            for (var i = 0; i < 20; i++)
            {
                values1[i] = random1.NextBoolean();
                values2[i] = random2.NextBoolean();
            }
            
            Assert.NotEqual(values1, values2);
        }

        [Fact]
        public void NextUInt64Test()
        {
            var lastValue = 0UL;
            var theSameCount = 0;
            for (var i = 0; i < 10; i++)
            {
                var newValue = RandomHelpers.Default.NextUInt64();
                if (newValue == lastValue)
                {
                    theSameCount++;
                }
                else
                {
                    lastValue = newValue;
                    theSameCount = 0;
                }
                Assert.InRange(RandomHelpers.Default.NextUInt64((0UL, 5UL)), 0UL, 5UL);
            }
            Assert.True(theSameCount < 8);
        }

        [Fact]
        public void NextBooleanTest()
        {
            var trueCount = 0;
            var falseCount = 0;
            for (var i = 0; i < 10; i++)
            {
                var newValue = RandomHelpers.Default.NextBoolean();
                if (newValue)
                {
                    trueCount++;
                }
                else
                {
                    falseCount++;
                }
            }
            Assert.True(trueCount > 0);
            Assert.True(falseCount > 0);
        }
    }
}

File: cpp/Platform.Random.Tests/RandomExtensionsTests.cpp
namespace Platform::Random::Tests
{
    TEST(RandomExtensionsTests, NextUInt64Test)
    {
        auto lastValue = 0UL;
        auto theSameCount = 0;
        for (auto i = 0; i < 10; i++)
        {
            auto newValue = NextUInt64(RandomHelpers::Default);
            if (newValue == lastValue)
            {
                theSameCount++;
            }
            else
            {
                lastValue = newValue;
                theSameCount = 0;
            }

            std::uint64_t temp = NextUInt64(RandomHelpers::Default, {0UL, 5UL});
            ASSERT_LE(temp, 5UL);
            ASSERT_GE(temp, 0UL);
        }
        ASSERT_LT(theSameCount, 8);
    }

    TEST(RandomExtensionsTests, NextUIntTemplateTest)
    {
        // Test NextUInt with different unsigned integer types
        auto value8 = NextUInt<std::uint8_t>(RandomHelpers::Default);
        auto value16 = NextUInt<std::uint16_t>(RandomHelpers::Default);
        auto value32 = NextUInt<std::uint32_t>(RandomHelpers::Default);
        auto value64 = NextUInt<std::uint64_t>(RandomHelpers::Default);
        
        // Test with ranges
        auto rangedValue8 = NextUInt<std::uint8_t>(RandomHelpers::Default, {0, 10});
        auto rangedValue16 = NextUInt<std::uint16_t>(RandomHelpers::Default, {0, 100});
        auto rangedValue32 = NextUInt<std::uint32_t>(RandomHelpers::Default, {0, 1000});
        auto rangedValue64 = NextUInt<std::uint64_t>(RandomHelpers::Default, {0UL, 10000UL});
        
        // Verify ranges
        ASSERT_LE(rangedValue8, 10);
        ASSERT_GE(rangedValue8, 0);
        ASSERT_LE(rangedValue16, 100);
        ASSERT_GE(rangedValue16, 0);
        ASSERT_LE(rangedValue32, 1000);
        ASSERT_GE(rangedValue32, 0);
        ASSERT_LE(rangedValue64, 10000UL);
        ASSERT_GE(rangedValue64, 0UL);
    }

    TEST(RandomExtensionsTests, NextBooleanTest)
    {
        auto trueCount = 0;
        auto falseCount = 0;
        for (auto i = 0; i < 10; i++)
        {
            auto newValue = NextBoolean(RandomHelpers::Default);
            if (newValue)
            {
                trueCount++;
            }
            else
            {
                falseCount++;
            }
        }
        ASSERT_GT(trueCount, 0);
        ASSERT_GT(falseCount, 0);
    }
}
